website/blog/testing.md

```diff
@@ -0,0 +1,604 @@
+            t.Fatal(err)
+        }
+    })
+}
+```
+
+The integration test lists all available products using the catalog service,
+adds one of every product to a cart using the cart service, and checks out using
+the checkout service. Service Weaver automatically initializes all these
+services, as well as the other services on which they depend (e.g., shipping
+service, email service, payment service, etc.). Moreover, Service Weaver reduces
+the tedium of writing integration tests by allowing you to use language native
+types and method calls to interact with various services.
+
+We can run this integration test as easily as running `go test`:
+
+```console
+$ go test
+PASS
+ok      github.com/ServiceWeaver/onlineboutique 0.026s
+```
+
+In addition to using the `weavertest.Local` runner, which runs all components
+locally in a single process, we can also use
+
+- the `weavertest.RPC` runner, which forces components to interact via RPC, and
+- the `weavertest.Multi` runner, which runs every component in a separate
+  OS process.
+
+Some components may be too slow or cumbersome to use in a test. In these cases,
+Service Weaver allows you to replace the component with a [**fake**][fakes]. For
+example, we can write a fake implementation of the product catalog service that
+uses a fixed in-memory set of products:
+
+```go
+type Product = productcatalogservice.Product
+
+type fakeCatalog struct {
+    products []Product
+}
+
+func (f *fakeCatalog) ListProducts(context.Context) ([]Product, error) {
+    return f.products, nil
+}
+
+func (f *fakeCatalog) GetProduct(ctx context.Context, id string) (Product, error) {
+    for _, p := range f.products {
+        if p.ID == id {
+            return p, nil
+        }
+    }
+    return Product{}, productcatalogservice.NotFoundError{}
+}
+
+func (f *fakeCatalog) SearchProducts(context.Context, string) ([]Product, error) {
+	panic("unimplemented")
+}
+```
+
+Then, we can update our integration test to use a `fakeCatalog` instead of the
+real product catalog service component.
+
+```go
+func TestPurchaseWithFakeCatalog(t *testing.T) {
+    catalog := &fakeCatalog{
+        products: []Product{
+            {
+                ID:          "0",
+                Name:        "Noogler Hat",
+                Description: "A colorful hat given to new Googlers.",
+                PriceUSD:    money.T{CurrencyCode: "USD", Units: 42},
+            },
+        },
+    }
+    fake := weavertest.Fake[productcatalogservice.ProductCatalogService](catalog)
+    runner := weavertest.Local
+    runner.Fakes = append(runner.Fakes, fake)
+    runner.Test(t, func(
+        t *testing.T,
+        catalog productcatalogservice.ProductCatalogService,
+        cart cartservice.CartService,
+        checkout checkoutservice.CheckoutService,
+    ) {
+        // Testing code goes here.
+        ...
+    })
+}
+```
+
+## 3. Randomized Testing
+
+When you write a unit test or integration test, you write some code and check
```

Comment:
Chatted offline: Maybe we can put all paras till "Service Weaver implements a type of randomized testing called ..." in a subsection called "Why Randomized Testing is Important?"

so we will have:

3. Randomized Testing
3.1 Why Randomized Testing is Important
3.2 Deterministic Simulation
   Move here the para "Service Weaver implements a type of randomized testing ..." and "To deterministically simulate a system without ..." 
 3.2.1 How to? or How to do Deterministic Simulation? or smthg like that   
  Here is the content of the deterministic simulation section as it is right now
  3.2.2 Example or Banking Example as you have it right now

Reply:
Done. Thanks Robert! It's much clearer now.